seed/unpack.js

```diff
@@ -0,0 +1,141 @@
+
+function expandedDescription(description) {
+  let out = [];
+  description.forEach(part => {
+    if (Array.isArray(part)) {
+      out.push('<div class="step">');
+
+      if (part[ 0 ]) {
+        out.push('<img src="' + part[ 0 ] + '">');
+      }
+      if (part[ 1 ]) {
+        out.push('<div class="caption">');
+        out.push(part[ 1 ]);
+        out.push('</div>');
+      }
+      if (part[ 2 ]) {
+        out.push('<div class="text">');
+        out.push(part[ 2 ]);
+        out.push('</div>');
+      }
+      if (part[ 3 ]) {
+        out.push('<div class="link">');
+        out.push('<a href="' + part[ 3 ] + '">');
+        out.push(part[ 3 ]);
+        out.push('</a></div>');
+      }
+      out.push('</div>');
+    } else {
+      out.push(part.toString());
+    }
+  });
+  // indent by 2
+  return out;
+}
+
+function expandedTests(tests) {
+  if (!tests) {
+    return [];
+  }
+  let out = [];
+  tests.forEach(test => {
+    if (_.isString(test)) {
+      out.push(test);
+    } else {
+      // todo: figure out what to do about these id-title challenge links
+      out.push('//' + JSON.stringify(test));
+    }
+  });
+  return out;
+}
+
+
+challenges.forEach(challengeGroup => {
+  let jsonFile = path.parse(challengeGroup.fileName);
+
+  let unpackedDir = path.join(
+    __dirname,
+    'unpacked',
+    jsonFile.dir,
+    jsonFile.name
+  );
+
+  fs.mkdirp(unpackedDir, (err) => {
+    if (err && err.code !== 'EEXIST') {
+      console.log(err);
+      throw err;
+    }
+
+    console.log('Unpacking "' + challengeGroup.name + '" into ' +
+      jsonFile.dir);
+
+    // todo: omit fileName, superBlock, superOrder extra properties
+    writeChallengeFile(JSON.stringify(challengeGroup, null, 2), unpackedDir,
+      jsonFile.name, '.json');
+
+    challengeGroup.challenges.forEach(challenge => {
+
+      // one file: description plus seed (HTML)
+      let text = [];
+      text.push('<html><body>');
+      // todo: CSS styles for previewing edits
+      text.push('<div class="description">');
+      text.push('<!--DESCRIPTION-->');
+      text.push(expandedDescription(challenge.description).join('\n'));
+      text.push('<!--END-->');
+      text.push('</div>');
+      text.push('');
+      text.push('<!--SEED--><pre>');
+      if (challenge.challengeSeed) {
+        text.push(text, challenge.challengeSeed.join('\n'));
+      }
+      text.push('<!--END-->');
+      text.push('</pre>');
+      text.push('</div></body></html>');
+      writeChallengeFile(text, unpackedDir, challenge.id, '.html');
+
+      // todo: unpack translations
+
+      // another file: solution plus tests (JavaScript)
+      // todo: use common/app/routes/Challenges/utils/index.js:15 maps
+      // to determine file extensions for non-JS challenges
+
+      text = [];
+      text.push('//--SOLUTION--');
+      // Note: none of the challenges have more than one solution
+      // todo: should we deal with multiple solutions or not?
```

Comment:
The `solution` is used only for CI. I couldn't see why we would have more than one

Reply:
at testfirst.org we allowed students to submit alternate solutions -- different algorithms or code styles can be illuminating... but having many solutions makes the repo harder to maintain, so probably this would be best implemented as a separate "submit your solution / see other solutions" feature to be done later